test_util/azure.py

```diff
@@ -0,0 +1,198 @@
+from msrestazure.azure_exceptions import CloudError
+from retrying import retry
+
+from test_util.helpers import lazy_property
+
+
+# Very noisy loggers that do not help much in debug mode
+logging.getLogger("msrest").setLevel(logging.INFO)
+logging.getLogger("requests_oauthlib").setLevel(logging.INFO)
+log = logging.getLogger(__name__)
+
+# This interface is designed to only use a single deployment
+# Being as the azure interface is based around resource groups, deriving
+# deployment name from group names makes it easier to attach to creating deploys
+DEPLOYMENT_NAME = '{}-Deployment'
+
+
+class AzureWrapper:
+    def __init__(self, subscription_id, tenant_id, client_id, client_secret):
+        self.credentials = azure.common.credentials.ServicePrincipalCredentials(
+            client_id=client_id,
+            secret=client_secret,
+            tenant=tenant_id)
+        self.rmc = ResourceManagementClient(self.credentials, subscription_id)
+        self.nmc = NetworkManagementClient(self.credentials, subscription_id)
+
+
+class TemplateDeployment:
+    def __init__(self, group_name, azure_wrapper):
+        # each template deployment is a resource group
+        self.group_name = group_name
+        self.azure_wrapper = azure_wrapper
+
+    @classmethod
+    def create(
+            cls, azure_wrapper, template_uri, location, group_name,
+            public_key, master_prefix, agent_prefix, admin_name, oauth_enabled,
+            vm_size, agent_count, name_suffix, vm_diagnostics_enabled):
+        """ Creates a new resource group and deploys the DC/OS template to it
+        """
+        assert master_prefix != agent_prefix, 'Master and agents must have unique prefixs'
+        assert re.match('^[a-z][a-z0-9-]{1,61}[a-z0-9]$', master_prefix),\
+            'Invalid master DNS prefix: {}'.format(master_prefix)
+        assert re.match('^[a-z][a-z0-9-]{1,61}[a-z0-9]$', agent_prefix),\
+            'Invalid agent DNS prefix: {}'.format(agent_prefix)
+
+        deployment_name = DEPLOYMENT_NAME.format(group_name)
+
+        # Resource group must be created before validation can occur
+        if azure_wrapper.rmc.resource_groups.check_existence(group_name):
+            raise Exception("Group name already exists / taken: {}".format(group_name))
+        log.info('Starting resource group_creation')
+        azure_wrapper.rmc.resource_groups.create_or_update(
+            group_name,
+            ResourceGroup(location=location))
+
+        template_params = {
+            'sshRSAPublicKey': public_key,
+            # must be lower or validation will fail
+            'masterEndpointDNSNamePrefix': master_prefix.lower(),
+            'agentEndpointDNSNamePrefix': agent_prefix.lower(),
+            'linuxAdminUsername': admin_name,
+            'agentVMSize': vm_size,
+            'agentCount': agent_count,
+            'nameSuffix': name_suffix,
+            'oauthEnabled': repr(oauth_enabled).lower(),
```

Comment:
we should put a type check on oauth_enabled. I'm guessing this makes it be `true`/`false` but just reading this code that isn't obvious